tasm/tiles/src/LayoutDatabase.cc

```diff
@@ -0,0 +1,386 @@
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 6, numCols));
+
+    int result = sqlite3_step(select);
+    ASSERT_SQLITE_DONE(result);
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+}
+
+void LayoutDatabase::addTileLayoutWidths(const std::string &video, unsigned int id, const std::vector<unsigned int> &widths) {
+    std::string query = "INSERT INTO widths (video, id, width, position) "\
+                    "VALUES ( ?, ?, ?, ?)";
+
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+
+    int position = 0;
+    for (const auto& width : widths) {
+        ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 3, width));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 4, position));
+        ASSERT_SQLITE_DONE(sqlite3_step(select));
+        ASSERT_SQLITE_OK(sqlite3_reset(select));
+        ++position;
+    }
+
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+}
+
+void LayoutDatabase::addTileLayoutHeights(const std::string &video, unsigned int id, const std::vector<unsigned int> &heights) {
+    std::string query = "INSERT INTO heights (video, id, height, position)"\
+                    "VALUES (?, ?, ?, ?)";
+
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+
+    int position = 0;
+    for (const auto& height : heights) {
+        ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 3, height));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 4, position));
+        ASSERT_SQLITE_DONE(sqlite3_step(select));
+        ASSERT_SQLITE_OK(sqlite3_reset(select));
+        ++position;
+    }
+
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+}
+
+unsigned int LayoutDatabase::totalHeight(const std::string &video) const {
+    // select all ids for this video from layouts
+    std::string query = "SELECT id FROM layouts WHERE video = ?";
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+    ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+
+    std::vector<unsigned int> ids;
+    int result;
+    while ((result = sqlite3_step(select)) == SQLITE_ROW) {
+        ids.push_back(sqlite3_column_int(select, 0));
+    }
+
+    ASSERT_SQLITE_DONE(result);
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+
+    // for all ids, check that the total height is the same
+    assert(ids.size() > 0);
+
+    std::vector<unsigned int> heights = heightsForId(video, ids.at(0));
+    unsigned int totalHeight = std::accumulate(heights.begin(), heights.end(), 0);
+    for (unsigned int id : ids) {
```

Comment:
If we don't verify that the height is the same for all IDs, then we could do this aggregation in the query: "SELECT sum(height) FROM heights WHERE video=? and id=?". On the other hand, it's nice to re-use the `heightsForId` function, despite it adding a bit to the complexity. It's up to you how you want to do this.